apps/teams-test-app/src/components/DialogAPIs.tsx

```diff
@@ -1,90 +1,125 @@
+          const childWindow = dialog.open(dialogInfo, onComplete);
+          childWindow.addEventListener('message', (message: string) => {
+            // Message from parent
+            setResult(message);
+          });
+          childWindowRef.current = childWindow;
+          return '';
+        },
+      },
+    });
 
-  const openDialog = (dialogInfoInput: string): void => {
-    const dialogInfo: DialogInfo = JSON.parse(dialogInfoInput);
-    const onComplete = (err: string, result: string): void => {
-      setOpenRes('Error: ' + err + '\nResult: ' + result);
-    };
-    setOpenRes('dialog.open' + noHostSdkMsg);
-    dialog.open(dialogInfo, onComplete);
-  };
+  const ResizeDialog = (): ReactElement =>
+    ApiWithTextInput<DialogInfo>({
+      name: 'dialogResize',
+      title: 'Dialog Resize',
+      onClick: async dialogInfo => {
+        dialog.resize(dialogInfo);
+        return 'Teams client SDK call dialog.resize was called';
+      },
+    });
 
-  const resizeDialog = (dialogInfoInput: string): void => {
-    const dialogInfo: DialogInfo = JSON.parse(dialogInfoInput);
-    dialog.resize(dialogInfo);
-    setResizeRes('Teams client SDK call dialog.resize was called');
-  };
+  const SendMessageToChild = (): ReactElement =>
+    ApiWithTextInput<string>({
+      name: 'sendMessageToChild',
+      title: 'sendMessageToChild',
+      onClick: async message => {
+        if (childWindowRef.current && childWindowRef.current !== null) {
+          const childWindow = childWindowRef.current;
+          await childWindow.postMessage(message);
+          return Promise.resolve('Message sent to child');
+        } else {
+          return Promise.resolve('childWindow doesnt exist');
+        }
+      },
+    });
 
-  const submitDialogWithInput = (submitDialogInput: string): void => {
-    if (submitDialogInput.length == 0) {
-      dialog.submit();
-      setSubmitRes('Teams client SDK call dialog.submit was called with no arguments');
-    } else {
-      try {
-        const parsedInput = JSON.parse(submitDialogInput);
-        dialog.submit(parsedInput.result, parsedInput.appIds);
-        setSubmitRes('Teams client SDK call dialog.submit was called with arguments');
-      } catch (error) {
-        if (error instanceof SyntaxError) {
-          setSubmitRes(generateJsonParseErrorMsg());
-        } else if (error instanceof Error) {
-          setSubmitRes(error.message);
+  const SendMessageToParent = (): ReactElement =>
+    ApiWithTextInput<string>({
+      name: 'sendMessageToParent',
+      title: 'sendMessageToParent',
+      onClick: async message => {
+        const parentWindow = ParentAppWindow.Instance;
+        if (parentWindow) {
+          await parentWindow.postMessage(message);
+          return Promise.resolve('Message sent to parent');
         } else {
-          setSubmitRes(JSON.stringify(error));
+          return Promise.resolve('parentWindow doesnt exist');
```

Comment:
because it's an async function, you shouldn't need to wrap your return in `Promise.resolve`

Reply:
The same should apply everywhere actually. You're using `Promise.resolve` in a lot of places where I don't think it's strictly necessary. Would you mind updating all these places with a simple `return '...'`?